python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -248,7 +282,9 @@ def _get_concurrent_resource_demand_to_launch(
 
             if nodes_to_add > 0:
                 updated_nodes_to_launch[node_type] = min(
-                    nodes_to_add, to_launch[node_type])
+                    max(nodes_to_add,
```

Comment:
There is an issue here that will require some thinking to resolve. We only get here if `nodes_to_add>0`, which for example, if we are starting to launch nodes and we currently have 5 pending and 0 running, we will never be `nodes_to_add>0` and hence if `nodes_to_add_based_on_requests.get(node_type, 0)=100`, it will not scale it.
If alternatively, you change it somehow, you need to make sure we do not end up with launching failing nodes continuously.

Reply:
I don't think this matters right? This is only about enforcing the rate limit. The calculation is done earlier already and if no nodes need to be added that's fine.

Reply:
Let me explain how this function works for us to be on the same page.
Example 1: Let's say we have 0 workers currently running, 0 pending launches, 0 pending nodes. We want to add another 100 because of `nodes_to_add_based_on_requests`. 
The functionality you describe will actually be the case and we will bypass the max concurrency launch (in that case `nodes_to_add = 5 > 0` and spin up 100 nodes.
Example 2: Let's say we have 0 workers currently running, 5 pending launches/nodes (the only difference from Example 1 is that we happen to have 5 pending nodes). We want to add another 100 because of `nodes_to_add_based_on_requests`. 
This _will not_ spin up the 100 nodes because at the current stage because `nodes_to_add = 5-5 = 0` and hence the if statement will be false.

I think this is not the behavior you expect. You would probably want to spin up the 100 nodes regardless of the pending nodes. When you do that (spin up 100 nodes regardless), you have to handle the case that the pending nodes can be failed nodes. I have the same issue when I want to perform aggressive autoscaling.

Reply:
> Example 1: Let's say we have 0 workers currently running, 0 pending launches, 0 pending nodes. We want to add another 100 because of nodes_to_add_based_on_requests.

I think you're confused here: we never add any nodes based on this dict. This dict is used *only for lifting the rate limiting restriction*. It is not used for adding nodes.

Reply:
I think the question we need to answer here is when do we lift the rate limit restriction? In the current situation, it only lifts it if `nodes_to_add` is bigger than 0, which means the lift happens only if the rate limit allows adding at least one node. Should this completely bypass the rate limit or somehow take it into account? 
Would it make sense to do:
`if nodes_to_add+nodes_to_add_based_on_requests.get(node_type, 0)>0`?